### PR TITLE
fix: handle inconsistent and empty lease specifications

### LINKED ISSUE #30: Lease left orphaned after rollout causes repeated 409 conflicts on renewTime
### Problem

After rollouts / on restart, HA components using this lease manager may fail to re-acquire a Kubernetes Lease, resulting in repeated 409 Conflict errors and blocking leader election.

### Cause

Lease object "corrupted" due to unknown reason.
May happen during rollout, Kubernetes maintenance or other pod rescheduling activities.

Example of state of the Lease: `renewTime` but no `holderIdentity` in the `spec`

`
apiVersion: coordination.k8s.io/v1
kind: Lease
metadata:
  creationTimestamp: "2025-10-14T06:18:01Z"
  name: test-lease
  namespace: test-namespace
  resourceVersion: "1768959024179407009"
  uid: 58dc9909-7227-4824-ab14-838d62a4fb08
spec:
  leaseDurationSeconds: 30
  leaseTransitions: 31
  renewTime: "2026-01-21T01:30:24.166773Z"
`

### Description

This issue has been occurring a couple of time over the last few months across multiple HA components using this crate. It seems to be happening during rollouts of new versions, although the exact cause is still under investigation. I suspect the lease being left in an incorrect state while the holder is being shutdown / other replicas are being started.

In affected cases, the Lease object is left orphaned. When replicas restart, they all attempt to re-acquire the same lease and consistently fail with a 409 Conflict on spec.renewTime (see log sample below).

My current assumption is that the lease is left in an invalid or “dirty” state when the holder is being shutdown and other replicas are starting. Possibly due to the patch method being used, Kubernetes no longer allows updates to the object. I am opening this issue to document the behavior and plan to work on a fix in the coming days. Any insights or recommendations would be greatly appreciated.

### Log sample

```
{"timestamp":"2026-01-21T00:20:50.311179Z","level":"TRACE","message":"try to lock orphaned lease","target":"kube_lease_manager::state","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311203Z","level":"DEBUG","message":"patch lease","patch":"Apply(Object {\"apiVersion\": String(\"coordination.k8s.io/v1\"), \"kind\": String(\"Lease\"), \"spec\": Object {\"acquireTime\": String(\"2026-01-21T00:20:50.311178Z\"), \"holderIdentity\": String(\"0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH\"), \"leaseDurationSeconds\": Number(30), \"renewTime\": String(\"2026-01-21T00:20:50.311178Z\")}})","target":"kube_lease_manager::state","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311250Z","level":"TRACE","message":"sending request to buffer worker","target":"tower::buffer::service","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311268Z","level":"TRACE","message":"worker polling for next message","target":"tower::buffer::worker","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311280Z","level":"TRACE","message":"processing new request","target":"tower::buffer::worker","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311289Z","level":"TRACE","resumed":false,"message":"worker received request; waiting for service readiness","target":"tower::buffer::worker","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311301Z","level":"DEBUG","service.ready":true,"message":"processing request","target":"tower::buffer::worker","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311316Z","level":"TRACE","message":"returning response future","target":"tower::buffer::worker","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311333Z","level":"TRACE","message":"worker polling for next message","target":"tower::buffer::worker","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311383Z","level":"DEBUG","message":"requesting","target":"kube_client::client::builder","span":{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"},"spans":[{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"}],"threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311420Z","level":"TRACE","message":"take? (\"https\", 10.76.0.1): expiration = Some(90s)","target":"hyper_util::client::legacy::pool","span":{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"},"spans":[{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"}],"threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.311452Z","level":"DEBUG","message":"reuse idle connection for (\"https\", 10.76.0.1)","target":"hyper_util::client::legacy::pool","span":{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"},"spans":[{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"}],"threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.320710Z","level":"TRACE","message":"put; add idle connection for (\"https\", 10.76.0.1)","target":"hyper_util::client::legacy::pool","span":{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"},"spans":[{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"}],"threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.320766Z","level":"DEBUG","message":"pooling idle connection for (\"https\", 10.76.0.1)","target":"hyper_util::client::legacy::pool","span":{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"},"spans":[{"http.method":"PATCH","http.url":"https://10.76.0.1/apis/coordination.k8s.io/v1/namespaces/drawer-workflows/leases/workflow-template-watchers?&fieldManager=kube-lease-manager-0ob4j5p3jcIz9JMUuf4wUnl3RFHNeTQH","otel.kind":"client","otel.name":"patch","name":"HTTP"}],"threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.320829Z","level":"DEBUG","message":"Unsuccessful: ErrorResponse { status: \"Failure\", message: \"Apply failed with 1 conflict: conflict with \\\"kube-lease-manager-Kq9VQryt5BWrQfMDyQwyEmVZUiR8da89\\\" using coordination.k8s.io/v1: .spec.renewTime\", reason: \"Conflict\", code: 409 }","target":"kube_client::client","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.320854Z","level":"DEBUG","message":"conflict detected while patching","error":"ErrorResponse { status: \"Failure\", message: \"Apply failed with 1 conflict: conflict with \\\"kube-lease-manager-Kq9VQryt5BWrQfMDyQwyEmVZUiR8da89\\\" using coordination.k8s.io/v1: .spec.renewTime\", reason: \"Conflict\", code: 409 }","target":"kube_lease_manager::state","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.320882Z","level":"TRACE","message":"generate next random duration","min":"2.500","val":"3.231","max":"5.000","target":"kube_lease_manager::backoff","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.320895Z","level":"TRACE","message":"next random duration requested","duration":"3.230892713152476","target":"kube_lease_manager::backoff","threadId":"ThreadId(2)"}
{"timestamp":"2026-01-21T00:20:50.320903Z","level":"TRACE","message":"backoff sleep","duration":"3.230892713s","target":"kube_lease_manager::backoff","threadId":"ThreadId(2)"}
```

### PR DESCRIPTION
**Pull request by nicolas-vivot**
Tentative PR to address this [issue](https://github.com/alex-karpenko/kube-lease-manager/issues/30)

Whenever a Lease exists with an inconsistent `spec` section, managers fails to acquire the lock even through nobody is actually holding it.

Inconsistent state for the specification is whenever either `renew_time` or `acquire_time` fields are present but `holder_identity` is missing.

### Tests

- existing unit tests
- added new unit tests to cover fixed edge cases

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -42,6 +42,8 @@ impl From<LeaseStateError> for LeaseManagerError {
             LeaseStateError::LeaseAlreadyExists(lease) => LeaseManagerError::LeaseAlreadyExists(lease),
             LeaseStateError::NonexistentLease(lease) => LeaseManagerError::NonexistentLease(lease),
             LeaseStateError::InconsistentState(err) => LeaseManagerError::InconsistentState(err),
+            LeaseStateError::InconsistentLease(_) => unreachable!("this branch is unreachable, looks like a BUG!"),
+            LeaseStateError::EmptyLeaseSpec(_) => unreachable!("this branch is unreachable, looks like a BUG!"),
```
Comment (nicolas-vivot) [inline src/error.rs]: Not a big fan but this is how it is done so far, i did not want to change your code style.

File: src/state.rs [hunk 2]
```diff
@@ -51,13 +51,28 @@ pub enum LeaseStateError {
...
     /// Internal lease state inconsistency detected.
     ///
     /// Usually the root cause is a bug.
     #[error("inconsistent LeaseManager state detected: {0}")]
     InconsistentState(String),
 }
 
+/// A deterministic name for unknown lease holder identity for inconsistent lease specifications.
+/// Note: Consider using random UUIDs to prevent collision with real identities.
+const INCONSISTENT_UNKNOWN_LEASE_HOLDER_IDENTITY: &str = "inconsistent_unknown_identity";
```
Comment (nicolas-vivot) [inline src/state.rs]: I did not want to introduce another crate just for this.

This should be good enough, considering that the only chance to get a collision is to override the default identity all along with this exact same value.

I leave it to you to decide whether or not to replace this with some random generation, using `uuid` or something else.

You may come with a slightly different idea for the fix that does not require this as well.

File: src/error.rs (hunk 1 above)
Comment (alex-karpenko) [reply src/error.rs]: It's because of `From<>` implementation on enum type. I'm not fan of generic (`_`) branch, preffering fine control on future changes of the source type. So it looks good in this particular case.

File: src/state.rs (hunk 2 above)
Comment (alex-karpenko) [reply src/state.rs]: This approach (setting of non-existent holder) looks hard but effective way to solve the issue. Let's try to test it. 

IMHO, using of something random would be beter to avoid potential (although almost non-realistic) conflicts. I suggest to use constant as prefix and add short (4-6 symbols) alphanumeric suffix (generated [like this](https://github.com/alex-karpenko/kube-lease-manager/blob/fe720f2266eef9e0acad6902d148a322b4c5c20e/src/manager.rs#L778-L781)). In such case we can see source/reason of this holder id, just for debug directly in cluster.

**Comment by alex-karpenko** [comment]
@nicolas-vivot, Thank you for contribution.
Could you please try to reproduce this case in the tests? If no, I can try.

File: src/state.rs (hunk 2 above)
Comment (nicolas-vivot) [reply src/state.rs]: ah! I knew i saw some random regeneration function somewhere but could not find it back.
Thank you, i will do that change after fully testing the solution with the different edge cases in local.

File: src/manager.rs [hunk 3]
```diff
@@ -775,7 +775,7 @@ fn random_duration(min_millis: DurationMillis, max_millis: DurationMillis) -> Du
     Duration::from_millis(rng().random_range(min_millis..=max_millis))
 }
 
-fn random_string(len: usize) -> String {
+pub(crate) fn random_string(len: usize) -> String {
```
Comment (nicolas-vivot) [inline src/manager.rs]: This was private and since the manager module is the one being exposed publicly i reduced the scope.

We have an import on `state` module, which then create a circular dependency between the two modules.

I am not sure how strict you are on limiting these things, do you want to re-organize code to avoid this, or you are good that way?

File: src/state.rs [hunk 4]
```diff
@@ -332,12 +380,47 @@ impl LeaseState {
...
     use crate::tests::{LeaseDropper, TEST_NAMESPACE, init, sleep_secs};
     use k8s_openapi::api::coordination::v1::LeaseSpec;
     use kube::api::DeleteParams;
 
-    async fn setup_simple_leaders_vec(lease_name: &str, count: usize) -> (Vec<LeaseParams>, Vec<LeaseState>) {
+    /// Set up a Lease with inconsistent spec for testing.
+    ///
+    /// Pass either `renew_time` or `acquire_time` as `Some` and the other as `None`.
+    /// Passing both as `Some` is also fine.
+    async fn setup_inconsistent_lease(
```
Comment (nicolas-vivot) [inline src/state.rs]: New test method to facilitate the creation of inconsistent lease objects.

File: src/state.rs [hunk 5]
```diff
@@ -701,4 +785,64 @@ mod tests {
 
         states[0].delete().await.unwrap();
     }
+
+    #[tokio::test]
+    #[ignore = "uses k8s current-context"]
+    async fn unattended_inconsistent_lease_soft_lock_release() {
```
Comment (nicolas-vivot) [inline src/state.rs:791]: New test to cover the edge cases that were failing before.

For each inconsistent combination:
- create the lease object with the inconsistent spec
- try to lock on all states
- verify that none of the state can actually acquire it
- verify that all states see the lease as locked and expired
- force release the lease (simulating what the manager would do in its next iteration
- try to lock on all states
- verify that state 0 acquired the lock

File: src/state.rs [hunk 6]
```diff
@@ -113,41 +127,67 @@ impl LeaseState {
...
+                // Inconsistent spec in the Lease - set holder to a deterministic unknown identity
+                // and enforce it as expired. This will allow the manager to proceed with releasing
+                // the lease.
+                Err(LeaseStateError::InconsistentLease(_)) => {
+                    let unknown_identity = format!("{}{}", INCONSISTENT_LEASE_HOLDER_IDENTITY_PREFIX, random_string(6));
+                    debug!(lease = %self.lease_name, "lease `spec` field is inconsistent. Considering it as held by {} and expired.", unknown_identity);
+                    self.holder = Some(unknown_identity);
+                    self.transitions = 0; // Should not matter
+                    self.expiry = SystemTime::now().checked_sub(Duration::from_nanos(1)).unwrap();
+                }
```
Comment (nicolas-vivot) [inline src/state.rs]: Whenever inconsistent Lease (e.g. `spec`) is detected, generate a random unknown identity and force the state to consider the lease locked by this identity and expired.

**Comment by nicolas-vivot** [comment]
The changes are bugged.

I did not realized that the method `get` on the state is being used in different places, so new errors need to be correctly handled in all places, or the manager fails to start with inconsistent lease.

I will fix it and add new tests at manager level to cover it.

File: src/state.rs [hunk 7]
```diff
@@ -388,13 +420,65 @@ mod tests {
...
+
+        let lease = state.create(LeaseCreateMode::UseExistent).await.unwrap();
+        assert!(lease.spec.is_some());
+        let spec = lease.spec.unwrap();
+        // Expect artificial holder identity and no renew time for immediate expiration
+        assert!(spec.holder_identity.is_some());
+        assert!(spec.renew_time.is_none());
+
+        state.delete().await.unwrap();
+    }
```
Comment (nicolas-vivot) [inline src/state.rs:473]: New tests to verify that inconsistent lease can be read when using AutoCreate or UseExistent modes.

File: src/manager.rs [hunk 8]
```diff
@@ -1585,4 +1624,28 @@ pub(crate) mod tests {
...
+        assert!(!manager0.is_leader.load(Ordering::Relaxed));
+
+        // Lock by 1st
+        let is_leader = manager0.changed().await.unwrap();
+        assert!(is_leader);
+        assert!(manager0.is_leader.load(Ordering::Relaxed));
+
+        // Clean up
+        manager0.state.read().await.delete().await.unwrap();
+    }
```
Comment (nicolas-vivot) [inline src/manager.rs:1653]: A new test to verify that the manager behave as expected with inconsistent lease.

The manager is expected to be able to read the lease, release it since it is considered as locked and expired and then acquire the lease.

File: src/state.rs [hunk 9]
```diff
@@ -271,7 +274,30 @@ impl LeaseState {
...
+                    spec.renew_time = None;
+                    lease.spec = Some(spec);
+
+                    Ok(lease)
+                }
+                // Consistent spec
+                Some(_) => Ok(lease),
+                // Empty spec
+                None => Ok(lease),
+            },
```
Comment (nicolas-vivot) [inline src/state.rs:300]: The previous approach (handling different type of errors) had a wide impact a bit everywhere.

To avoid to refactor too many parts to handle errors consistently, i decided to opt for this approach: modify the state of `spec` on the fly.

**Comment by nicolas-vivot** [comment]
@alex-karpenko This is the final version.

I have changed for a different approach as commented in the code, to limit the refactoring / impacts.
I added tests to cover the new behavior at creation and runtime for both state and manager.
I have also redeployed my services in a development environment with this patch, running through all the known edge cases and made sure everything works with no regression.

No issue found so far.

File: src/manager.rs [hunk 10]
```diff
@@ -857,6 +864,36 @@ pub(crate) mod tests {
...
+        acquire_time: Option<MicroTime>,
+    ) -> Result<(), LeaseStateError> {
+        let client = init().await;
+        let api = Api::<Lease>::namespaced(client, TEST_NAMESPACE);
+
+        let pp = PostParams::default();
+        // Creates a Lease without holder identify but with renew_time/acquire_time
+        let mut spec = LeaseSpec::default();
+        spec.renew_time = renew_time;
+        spec.acquire_time = acquire_time;
```
Comment (alex-karpenko) [inline src/manager.rs]:
```suggestion
        let spec = LeaseSpec {
            renew_time,
            acquire_time,
            ..Default::default()
        };

```

**Comment by alex-karpenko** [comment]
@nicolas-vivot  I added tiny suggestion to fix linter finding.
Everything else looks good to me, thank you a lot!

**Comment by nicolas-vivot** [comment]
No rush, do you think we can have a new release soon?
Just to know if it is worth on my side to create a tag on my fork just for the sake to push this to production for my company's products.

If you intend to release soon after merging this, i will probably just wait for it since i already have counter measures to handle this issue on production when it appears.

**Review by alex-karpenko — APPROVED** [review]

**Comment by alex-karpenko** [comment]
It'll be merged today and v0.11.1 will be released just after that.

**Comment by alex-karpenko** [comment]
@nicolas-vivot release 0.11.1 has been published, thank you for efforts.
